### DISCUSSION TITLE
Publish/subscrive SSE mechanism performance with a lot of publishers

### DISCUSSION THREAD

**Discussion by mlusetti** [pocketbase/pocketbase#6168] [Q&A] [answered]
Hi,
  I'm the situation where I will have around 55K devices calling a custom API which will programmatically send messages to subscribed clients.

Clients will be a lot less, the maximum will probably be around 10 (ten).

Data will not be serialized into DB unless there's a specific type of message sent.
All data will be send to all clients connected.

I've tested this mechanism with 200 devices and is working smoothly.

Do you see any issues in this architecture ?
Thanks

**Comment by ganigeorgiev** [comment, marked as answer]
It is difficult to say as it will depend on the the messages payload and the available resources (primarily RAM, CPU, max allowed file descriptors).

If there are no access checks needed and you are not persisting most of the messages in the database, then I don't expect PocketBase to be much of a bottleneck because essentially you are only json marshal/unmarshal some data and sending the result over a channel.

From my experience, the most common error when having many realtime connections or concurrent requests is reaching the file descriptors limit (see https://pocketbase.io/docs/going-to-production/#increase-the-open-file-descriptors-limit), so make sure that either your user running the process is not restricted or have a reasonably large limit set to accommodate your use case (_usually should be something like "max expected concurrent requests/connections + ~150 for the db connections + some small additional buffer"_). _But in your case with ~10 clients this shouldn't really be an issue._

@benallfree mentioned recently that one of his users got viral with their app and had performance issues with the realtime connections and disabling them somehow resolved it, but there are too much unknown details to determine what could have been the cause for it (_especially if there is a reverse proxy and hundreds of instances on the same machine competing for resources_).

**Reply by benallfree** [reply]
> especially if there is a reverse proxy and hundreds of instances on the same machine competing for resources

We considered this possibility, but other instances running at the time were unaffected. A hook file change to trigger a PB restart would cure the condition temporarily until the clients were reconnected.

We also adjusted db indexes.

There certainly could be other factors though. Once PocketHost has the ability to isolate instances on their own machines we may be able to understand more.

**Comment by mlusetti** [comment]
Thank for your answer,  is more then enough to convince me to follow this path.
Special thanks for your advice on fd limits... will keep that in mind!

I'll wait for Ben to comment if he thinks is needed.

**Reply by benallfree** [reply]
It sounds like you may have thousands of devices reporting live updates, but only a few hundred clients monitoring those updates via the realtime SSE, is that right?

If so, you should have no trouble. 

If you mean 55k realtime connections I would definitely monitor as things scale. One scaling option would be to build some “relay” servers that listen for real-time events from the origin and relay them to connected clients. A mini real-time mirror. We are using this approach in PocketHost and it works well.

**Reply by mlusetti** [reply]
> It sounds like you may have thousands of devices reporting live updates, but only a few hundred clients monitoring those updates via the realtime SSE, is that right?

Yeah, I'll have far less then few hundred... As I said there will be around 10 clients.

 > If so, you should have no trouble.
 
Good to know.

> If you mean 55k realtime connections I would definitely monitor as things scale. One scaling option would be to build some “relay” servers that listen for real-time events from the origin and relay them to connected clients. A mini real-time mirror. We are using this approach in PocketHost and it works well.

Nope I'll have 55K publishers (devices call HTTP endpoints) and have around 10 web clients on the realtime feature.

Thanks for your hints!

**Reply by benallfree** [reply]
>  will have around 55K devices calling a custom API which will programmatically send messages to subscribed clients

Can you say more about this setup? Do you mean the 55k devices will ping PocketBase with record updates and the 10 clients will be listening/subscribed via PocketBase's realtime SSE feature? 

If you mean the devices will be doing upserts to PocketBase while the 10 clients listen, the frequency of those 55k updates will have a large impact and could produce significant realtime chatter. Suppose each device sends and update once per minute. So 55k updates per minute, and 10 clients are listening. That means PocketBase will be pushing about 10k realtime updates per second.

**Reply by mlusetti** [reply]
Our devices will "ping" our custom PB instance around every 10 minutes (average) so we will have 6K HTTP post every 60 seconds (on average).
Which means 100 req/s circa. There will be spike since the 10 minutes window is not equally distributed but they will be not so heavy (devices talks to each other too and can orchestrate this comms).
Sometimes, not so frequently, will come in data that need to be stored in sqlite.
Average JSON playload for device ping is 1.5k but not all that will be relayed through PB realtime to clients, just a small part of it.

On the other side we have 10 (more or less) javascript clients subscribed to realtime updates (from PB).

There's an idea to cut off device update frequency in half so to double the req/s rate, but for now these are the coarse numbers.

**Reply by benallfree** [reply]
Thanks for sharing that. So it sounds like you may have as many as 1k/s realtime updates going out, assuming 10 realtime clients. 

You'll have to see how that goes. If you don't need every single realtime update going to every client all the time, there may be further ways to optimize. But in general, the advice is to build until it breaks and then refactor/optimize :)

**Reply by sksar** [reply]
Hey @mlusetti, 100 requests per second isn’t a big deal for Pocketbase. Even with spikes, it can handle way more easily on a small VM. You can check out the Pocketbase benchmarks here: https://github.com/pocketbase/benchmarks.

And the real-time connections to the 10 clients, listening for these events, won’t be a problem either. Pocketbase is built to handle these kinds of uses.

There’s no need to worry about premature optimization, as @benallfree said, build until it breaks. If you need to scale, you can just increase the VM and maybe do some tuning to your Reverse Proxy, like Nginx or Caddy.
